### ISSUE TITLE
UIkit 4: Shorten Class Names?

### ISSUE THREAD

**Issue by saschadube** [uikit/uikit#3984]
Should UIkit 4 shorten class names?

uk-background-* → uk-bg-*
uk-margin-left-* → uk-ml-*
uk-padding-x-* → uk-px-*
uk-width-* → uk-w-*
uk-height-* → uk-h-*

**Comment by yaneony** [comment]
Definitely a "NO". How would it look with "uk-margin-remove-last-child" like? "uk-mg-rm-lt-ch"? Sorry, not readable at all...

**Comment by htkm** [comment]
I am not sure if shortening names would be a good idea.

**Comment by saschadube** [comment]
@yaneony In UIkit 4 the long class name would be `uk-margin-last-child-none` and the short version could be `uk-mlc-none`

**Comment by yaneony** [comment]
@saschadube using ONLY short version, like i've said before, will lead new users to nowhere without searching whole documentation. Adding both variants, long and short would work well, but it will blow the framework to few MB...

```
.uk-margin-last-child-none, .uk-mlc-none {
  ...
}
```

**Comment by ryhnkml** [comment]
Keep them long

**Comment by hebinet** [comment]
I am a big fan of the short syntax!

I am using bootstrap and tailwind lately quite often and it is so convenient having short class names.
Ok, at first it was a bit weird but after a short period of time it was the best thing ever.

Lets do a small margin right :
UIkit: `uk-margin-small-right` => Tailwind: `mr-1`

or let do a large padding on the x axis:
UIkit: `uk-padding-large-left uk-padding-large-right` => Bootstrap: `px-5`

Ok, UIkit has no special classes for x/y axis margins or paddings but it is so intuitive to right the short syntax.

Lately asset size is also a big thing and could be shortening the classnames also decrease the resulting css size or even the resulting html size?

I quite often use margin or padding classes in my applications and websites. Maybe this could decrease loading time at least a bit.

I would recomend the following syntax (copied from the bootstrap documentation and addapted for uikit ;) ):

> The classes are named using the format `{property}{sides}-{size}` for all devices and `{property}{sides}-{size}@{breakpoint}` for s, m, l, and xl breakpoints.
> 
> Where property is one of:
> `m` - for classes that set margin
> `p` - for classes that set padding
> 
> Where sides is one of:
> `t` - for classes that set margin-top or padding-top
> `b` - for classes that set margin-bottom or padding-bottom
> `l` - for classes that set margin-left or padding-left
> `r` - for classes that set margin-right or padding-right
> `x` - for classes that set both *-left and *-right
> `y` - for classes that set both *-top and *-bottom
> `blank` - for classes that set a margin or padding on all 4 sides of the element
> 
> Where size is one of:
> `0` ore `none` - equal to -remove
> `s` - equal to -small
> `m` - equal to -medium
> `l` - equal to -large
> `xl` - equal to -xlarge
> `xxl` or `2xl` - equal to -xxlarge
> `auto` - for classes that set the margin to auto

**Comment by progreccor** [comment]
kepp them long. Don't shorten names.

**Comment by Oudmane** [comment]
Can we have them both ? Like we can set/change the prefix uk- , can we have an option for short or long before we build ?

**Comment by Septdir** [comment]
Not necessary. Now the classes, although long, are easy to remember and logical.

**Comment by ThomasMaitret** [comment]
Can shortening class names affect performance ?

**Comment by saschadube** [comment]
@ThomasMaitret No, it doesn't affect performance.

**Comment by htkm** [comment]
I am with people who think the syntax could feel pretty much long. But also who think directives like ‘mlc’ could look like another learning curve.

i have feeling that uikit is excellent when combining classnames derived from other categories(components).

even tough this a sort of advanced use cases, inferring short class names mixed and attached to an element could give a slight headache.

having them in both feels like a good approach. how much does it have to sacrifice the library size?

anyways, i am lean to those long names.

**Comment by maxsite** [comment]
My version: https://github.com/maxsite/berry

**Comment by Hurmeli** [comment]
I would vote for keeping the long names. It keeps the code readable and it's also easier to remember the class names.

**Comment by josepuma** [comment]
Explicit is always better than implicit.

**Comment by staticwave** [comment]
If you're going to adopt short names, I think the long names should still remain. One of the things I appreciate about Uikit is the verbose class names even after using it for many years.

**Comment by impranay** [comment]
Or can we have both release option, where we can choose, which distribution we want to go with..

**Comment by joeyslack** [comment]
There's no reason that the "short names" idea can't just be an extension library instead.

Keep the core working with long names, and make an official "short names" extension, as a sass/less mixin... problem solved. Just a simple mapping gulp task. I'd be happy to make this.

**Comment by Grawl** [comment]
maybe, atomic CSS is not such a good idea?

**Comment by hebinet** [comment]
I don't think that it is necessary to shorten every aspect of the framework but classes which are used very often like padding, margin or width should be shortened.

When I am developing a site I constantly swap margin or padding classes to get the best looking result possible.
In the IDE this is not such a big deal because it there is typehinting but when I use the Chrome Dev Tools I have to write the complete words over and over again.

So shortnames would also speedup the development process. Just swap one or two letters and the new result is visible. It was never that easy :)

**Comment by Hurmeli** [comment]
Whether you end up using long or short names, I really hope you keep the uk- prefix. It makes it possible to use uikit in conjuction with other frameworks.

**Comment by Taimurian** [comment]
Personally, I prefer the long names but can you have the short names option in the build process.

**Comment by sudoranger** [comment]
I definitely prefer long names for readability.

**Comment by wolfcreative** [comment]
One of the things I love about UIkit is the long class names. You ALWAYS know what this or that class will be doing. There is no confusion. The code is beautiful and concise.

**Comment by yazzbe** [comment]
small, medium, extra large ... those can be short. but keep the rest long and descriptive.

but please create alternatives (0-9, a-z) for responsive classes; I use uikit3 in Drupal, but many UI modules filter out the @ sign. Therefore I've created a file with alternative responsive classes: eg. 'uk-text-center@m' becomes 'uk-text-center--m'

**Comment by Grawl** [comment]
`@` is for `at`

**Comment by tmt307** [comment]
I think classes need to stay the same way they are so then it doesn't get confusing and so there isn't something new to memorize...

**Comment by josephsimony** [comment]
Not sure is this topic useful for an actual decision.

Anyways; if it's a vote: 
**Please keep it long or let us use both** varieties  (which complicates things).

Thanks :)

**Comment by Jdeboysere** [comment]
1. Don't shorten name
2. Use the last feature of SASS (@use, ...) : https://css-tricks.com/introducing-sass-modules/
8. `We remove redundancy into each set of unit classes, so that for instance instead of .uk-width-3-6 you should use .uk-width-1-2.` is REALLY stupid. Add .uk-width-3-6 and its friends, PLEASE. **Make our work simpler !**
3. More responsive classes : we have `.uk-text-center` or `.uk-text-left@m` and it really help. But a lot of useful classes dont have this flexibility !
4. Why a margin-bottom on the Subnav component ?!
5. Remove margin-bottom on the last child element and remove the margin-top on the first child element.
6. Remove useless component/classes/variables. There are TOO many variabes!
7. Be able to display the offcanvas from the top
9. Add a Datepicker. Many many (many...) clients don't like native Datepicker (too ugly, not enough customisable).
10. Use the same HTML structure for the Slider/Slidershow component with or without navigation. 
11. Be able to disable the auto-upload for the Upload component
12. The Lightbox component should not require a container. We should be able to apply on it directly on a link/img.
13. The Image component should be simpler and easy to implement.
14. Add a lazyload with blur effect, like this: https://jmperezperez.com/medium-image-progressive-loading-placeholder/. It's so beautiful.
15. The uk-grid component should not add the .uk-grid class. This is not a good practice because the grid can make a while to be built and this is really ugly.

In four words : Make it simpler, please.

**Comment by Grawl** [comment]
I think comment from @Jdeboysere from above is for #3967

**Comment by saschadube** [comment]
@Jdeboysere @szabeszg At the moment it's possible to create some kind of preload state with CSS while the image is loading. 

``` css
/* Example rule to apply a background while loading */
`img[data-src*='.jpg'][src*='data:image'] {
    background: grey;
}`
```

**Comment by saschadube** [comment]
@Grawl Ah, that's right 😅. @Jdeboysere Please move your comment over there https://github.com/uikit/uikit/issues/3967

**Comment by sudoranger** [comment]
> small, medium, extra large ... those can be short. but keep the rest long and descriptive.
> 
> but please create alternatives (0-9, a-z) for responsive classes; I use uikit3 in Drupal, but many UI modules filter out the @ sign. Therefore I've created a file with alternative responsive classes: eg. 'uk-text-center@m' becomes 'uk-text-center--m'

This should be at https://www.drupal.org/project/issues/drupal?categories=All instead of here. Uikit3 works very well with **WordPress** and no conflict whatsoever with **Ghost** and other top software publishers. This is definitely a Drupal flaw. Feel free to lodge a complain there to add to the _21k+ issues currently open_. Changing @ to -- will have some compatibility issues (if removed) and needs extra attentions/supports from the theme authors in the future and that sucks if we get paid so little to cover our cost/efforts.

TL;DR: If it ain't broke, don't fix it.

**Comment by ketan808** [comment]
Keep previous long and also have short class names.

**Comment by dertuerke** [comment]
> @ThomasMaitret No, it doesn't affect performance.

Of course it has an effect on it.
Every character that can be saved means 1 byte less that needs to be transferred!

Here is a small example based on @hebinet proposal:

```css
/*
  Original Code
  Size: 191 bytes
  Gzip: 94 bytes
*/
.uk-margin-small-right {
  margin-right: 16px;
}
.uk-margin-small-left {
  margin-left: 16px;
}
.uk-margin-small-top {
  margin-top: 16px;
}
.uk-margin-small-bottom {
  margin-bottom: 16px;
}

/*
  Minified Code
  Size: 160 bytes
  Gzip: 85 bytes
*/
.uk-margin-small-right{margin-right:16px}.uk-margin-small-left{margin-left:16px}.uk-margin-small-top{margin-top:16px}.uk-margin-small-bottom{margin-bottom:16px}

/*
  Original Code with shorthand
  Size: 137 bytes
  Gzip: 85 bytes
*/
.uk-mr-s {
  margin-right: 16px;
}
.uk-ml-s {
  margin-left: 16px;
}
.uk-mt-s {
  margin-top: 16px;
}
.uk-mb-s {
  margin-bottom: 16px;
}

/*
  Minified Code
  Size: 106 bytes
  Gzip: 76 bytes
*/
.uk-mr-s{margin-right:16px}.uk-ml-s{margin-left:16px}.uk-mt-s{margin-top:16px}.uk-mb-s{margin-bottom:16px}
```

Now this is just a small example. But we see that we save 9 bytes with GZIP.

**Comment by Grawl** [comment]
atomic CSS is a lie

**Comment by TimVanHerwijnen** [comment]
@saschadube 
The opinions on your suggestion are sure divided, but it's not a crazy idea. I suggest making it a fork or separate version.

**Comment by zzseba78** [comment]
Of course! do it! I will go a bit more deep:

- **Simpler, shorter and more intuitive markup in css classes.** - Speed up the forkflow!
So, instead of 
`uk-button uk-button-primary uk-button-small uk-border-rounded uk-margin-top-small uk-width-1-2` 
this:
 `btn is-primary is-small is-rounded mt-sm w-1/2`
- Work with a complete icon library - UIkit icons are great, but doesn´t cover basic utilities, and most of the time get mixed by uikit icons and other libraries to complement. Go for a one robust icon library, or let select one of choice.
- **Chante size suffix to prefixes**: Clearer and more simple markup ( like Tailwind does ), for example instead of:
`uk-width-1-2 uk-width-auto@s uk-width-1-1@m uk-width-expand@l`
  this:
`w-1/2 sm:w-auto md:w-full lg:w-expand`

**Comment by bzupancic** [comment]
I have always liked the explicit nature or UiKit, there is never confusion or interpretation needed - what you see is what you get.

A quick and easy win would be the removal of "uk-". Please, please, please let us remove the prefix if we don't need it for scoping!

Just spit-balling here, but a simple solution may be to simply let users decide within a .scss configuration file:

`
$prefix: 'uk-' !default;

$margin: 'margin' !default;
$margin-top: 'margin-top' !default;
$margin-right: 'margin-right' !default;
$margin-bottom: 'margin-bottom' !default;
$margin-left: 'margin-left' $default;

$xsmall: 'xsmall' !default;
$small: 'small' !default;
$medium: 'medium' !default;
$large: 'large' !default;
xlarge: 'xlarge' !default;

etc...
`

Could be modified based on the developers preference, like this:

`
$prefix: '' !default;

$margin: 'm' !default;
$margin-top: 'mt' !default;
$margin-right: 'mr' !default;
$margin-bottom: 'mb' !default;
$margin-left: 'ml' $default;

$xsmall: '-xs' !default;
$small: '-sm' !default;
$medium: '-md' !default;
$large: '-lg' !default;
$xlarge: '-xl' !default;

etc...
`

`
.#{prefix}#{margin-left}-#{large} {
     margin: 40px
}
`
Would translate to .uk-margin-left-large { 40px } or ml--lg { 40px }


*notice the BEM-like addition of the "--" on the modifier and the inclusion of the dash inside the prefix. :)

**Comment by dmleeman** [comment]
Long.

Heart says: I don't like *anything* Bootstrap! (yuk)

Head says: long names are their own cheat sheet.

**Comment by jonasmarco** [comment]
please use classes .row and .col for grid like the other frameworks

**Comment by prfardin** [comment]
> @ThomasMaitret No, it doesn't affect performance.

it's does not effect css file size even?
